modules/generator/registry/histogram.go

```diff
@@ -0,0 +1,207 @@
+package registry
+
+import (
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/storage"
+	"go.uber.org/atomic"
+)
+
+type histogram struct {
+	nameCount    string
+	nameSum      string
+	nameBucket   string
+	labels       []string
+	buckets      []float64
+	bucketLabels []string
+
+	// seriesMtx is used to sync modifications to the map, not to the data in series
+	seriesMtx sync.RWMutex
+	series    map[uint64]*histogramSeries
+
+	onAddSerie    func(count uint32) bool
+	onRemoveSerie func(count uint32)
+}
+
+type histogramSeries struct {
+	// labelValues should not be modified after creation
+	labelValues []string
+	count       *atomic.Float64
+	sum         *atomic.Float64
+	buckets     []*atomic.Float64
+	bucketInf   *atomic.Float64
+	lastUpdated *atomic.Int64
+}
+
+var _ Histogram = (*histogram)(nil)
+var _ metric = (*histogram)(nil)
+
+func newHistogram(name string, labels []string, buckets []float64) *histogram {
+	bucketLabels := make([]string, len(buckets))
+	for i, bucket := range buckets {
+		bucketLabels[i] = formatFloat(bucket)
+	}
+
+	return &histogram{
+		nameCount:    fmt.Sprintf("%s_count", name),
+		nameSum:      fmt.Sprintf("%s_sum", name),
+		nameBucket:   fmt.Sprintf("%s_bucket", name),
+		labels:       labels,
+		buckets:      buckets,
+		bucketLabels: bucketLabels,
+		series:       make(map[uint64]*histogramSeries),
+	}
+}
+
+func (h *histogram) setCallbacks(onAddSeries func(count uint32) bool, onRemoveSeries func(count uint32)) {
+	h.onAddSerie = onAddSeries
+	h.onRemoveSerie = onRemoveSeries
+}
+
+func (h *histogram) Observe(labelValues *LabelValues, value float64) {
+	if len(h.labels) != len(labelValues.getValues()) {
+		panic(fmt.Sprintf("length of given label values does not match with labels, labels: %v, label values: %v", h.labels, labelValues))
+	}
+
+	hash := labelValues.getHash()
+
+	h.seriesMtx.RLock()
+	s, ok := h.series[hash]
+	h.seriesMtx.RUnlock()
+
+	if ok {
+		h.updateSeries(s, value)
+		return
+	}
+
+	if !h.onAddSerie(h.activeSeriesPerHistogramSerie()) {
```

Comment:
Maybe a nil check here?

Reply:
I do a nil check in `newCounter`/`newHistogram` now.